docs/extend/middleware.md

```diff
@@ -1 +1,108 @@
-#  Middleware
\ No newline at end of file
+#  Middleware
+
+Middleware is a nifty way to filter HTTP requests in Flarum. This can allow you to, for example, run a check against an external database checking a user when they register, either allowing them to continue the process or not. The possibilities are endless!
+
+## Adding Middleware In Your Extension
+
+To add a new middleware, add a listener in your `extend.php` file:
+
+```php
+$events->subscribe(Listeners\AddMiddleware::class);
+```
+
+Now, in your listener, you can define the middleware:
+
+```php
+use Flarum\Event\ConfigureMiddleware;
+use Illuminate\Contracts\Events\Dispatcher;
+use YourMiddleware;
+
+class AddMiddleware
+{
+    public function subscribe(Dispatcher $events)
+    {
+        $events->listen(ConfigureMiddleware::class, [$this, 'addMiddleware']);
+    }
+    
+    public function addMiddleware(ConfigureMiddleware $event)
+    {
+        $event->pipe(app(YourMiddleware::class));
+    }
+}
+```
+
+Tada! Middleware defined. Now onto the middleware itself:
+
+```php
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+class YourMiddleware implements MiddlewareInterface {
+  public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+  {
+    // Your logic here!
+  }
+}
+```
```

Comment:
I would suggest moving this middleware logic (with at least a `return $handler->handle($request)`) to the top, right below the intro paragraph. For somebody who does not know what middleware does, seeing one is probably more important than learning how to register them with Flarum.